// semforms/lib/components/GeneratedStep.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require("babel-runtime/core-js/object/get-prototype-of");

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require("babel-runtime/helpers/possibleConstructorReturn");

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require("babel-runtime/helpers/inherits");

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _objectAssign = require("object-assign");

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _Question = require("./Question");

var _Question2 = _interopRequireDefault(_Question);

var _Configuration = require("../model/Configuration");

var _Configuration2 = _interopRequireDefault(_Configuration);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GeneratedStep = function (_React$Component) {
    (0, _inherits3.default)(GeneratedStep, _React$Component);

    function GeneratedStep(props) {
        (0, _classCallCheck3.default)(this, GeneratedStep);

        var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(GeneratedStep).call(this, props));

        _this._onStoreTrigger = function () {
            _this.setState({ question: _Configuration2.default.wizardStore.getStepData(_this.props.stepIndex) });
        };

        _this.onChange = function (index, change) {
            _Configuration2.default.wizardStore.updateStepData(_this.props.stepIndex, (0, _objectAssign2.default)(_this.state.question, change));
        };

        _this.state = {
            question: _Configuration2.default.wizardStore.getStepData([_this.props.stepIndex])
        };
        return _this;
    }

    (0, _createClass3.default)(GeneratedStep, [{
        key: "componentDidMount",
        value: function componentDidMount() {
            this.unsubscribe = _Configuration2.default.wizardStore.listen(this._onStoreTrigger);
        }
    }, {
        key: "componentWillUnmount",
        value: function componentWillUnmount() {
            this.unsubscribe();
        }
    }, {
        key: "render",
        value: function render() {
            return _react2.default.createElement(_Question2.default, { question: this.state.question, onChange: this.onChange, withoutPanel: true });
        }
    }]);
    return GeneratedStep;
}(_react2.default.Component);

GeneratedStep.propTypes = {
    stepIndex: _react2.default.PropTypes.number.isRequired
};
exports.default = GeneratedStep;
